cvmfs/export_plugin/posix/interface.cc

```diff
@@ -0,0 +1,743 @@
+    return -1;
+  }
+  return 0;
+}
+
+void posix_do_ffree(void *file_ctx) {
+  struct posix_file_handle *handle =
+    reinterpret_cast<posix_file_handle *>(file_ctx);
+  if (handle->fd != NULL) {
+    posix_do_fclose(file_ctx);
+  }
+  delete handle;
+}
+
+int posix_garbage_collector(struct fs_traversal_context *ctx) {
+  return RunGarbageCollection(ctx);
+}
+
+bool posix_archive_config(
+  std::string config_name,
+  std::string prov_name
+) {
+  int retval;
+  FILE *config = fopen(config_name.c_str(), "r");
+  if (config == NULL) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+      "Failed to open : %s : %d : %s\n",
+      config_name.c_str(), errno, strerror(errno));
+    return 1;
+  }
+
+  FILE *prov = fopen(prov_name.c_str(), "w");
+  if (prov == NULL) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+      "Failed to open : %s : %d : %s\n",
+      prov_name.c_str(), errno, strerror(errno));
+    return 1;
+  }
+
+  size_t bytes_transferred = 0;
+  while (1) {
+    char buffer[COPY_BUFFER_SIZE];
+
+    size_t actual_read = 0;
+    actual_read = fread(&buffer, sizeof(char), sizeof(buffer), config);
+    if (actual_read < sizeof(buffer) && ferror(config) != 0) {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+        "Read failed : %d %s\n", errno, strerror(errno));
+      return false;
+    }
+    bytes_transferred+=actual_read;
+    size_t written_len = fwrite(buffer, sizeof(char), actual_read, prov);
+    if (written_len != actual_read) {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+        "Write failed : %d %s\n", errno, strerror(errno));
+      return false;
+    }
+
+    if (actual_read < COPY_BUFFER_SIZE) {
+      break;
+    }
+  }
+
+  retval = fclose(config);
+  if (retval != 0) {
+    // Handle error
+    LogCvmfs(kLogCvmfs, kLogStderr,
+      "Failed close config file : %s : %d : %s\n",
+      config_name.c_str(), errno, strerror(errno));
+    return false;
+  }
+
+  retval = fclose(prov);
+  if (retval != 0) {
+    // Handle error
+    LogCvmfs(kLogCvmfs, kLogStderr,
+      "Failed close provenance file : %s : %d : %s\n",
+      prov_name.c_str(), errno, strerror(errno));
+    return false;
+  }
+
+  return true;
+}
+
+void posix_write_provenance_info(
+  struct fs_traversal_context *ctx,
+  const char *info_file
+) {
+  FILE *f = fopen(info_file, "w");
+  if (f != NULL) {
+    fwrite("repo : ", sizeof(char), 7, f);
+    fwrite(ctx->repo, sizeof(char), strlen(ctx->repo), f);
+    fwrite("\nversion : ", sizeof(char), 11, f);
+    fwrite(ctx->lib_version, sizeof(char), strlen(ctx->lib_version), f);
+    fwrite("\nbase : ", sizeof(char), 8, f);
+    if (ctx->base)
+      fwrite(ctx->base, sizeof(char), strlen(ctx->base), f);
+    fwrite("\ncache : ", sizeof(char), 9, f);
+    if (ctx->data)
+      fwrite(ctx->data, sizeof(char), strlen(ctx->data), f);
+    fwrite("\nconfig : ", sizeof(char), 10, f);
+    if (ctx->config)
+      fwrite(ctx->config, sizeof(char), strlen(ctx->config), f);
+    fclose(f);
+  }
+}
+
+void posix_archive_provenance(
```

Comment:
We should review this implementation. I think it is quite fragile and limits how we add information from the source repo.